### PR TITLE
feat: add password reset process

### PR DESCRIPTION
**Pull request by Maaayhan**
### Task description:

Added the process of resetting user password:
 - implemented mutation 'forgotPassword' and 'resetPassword'
 - refactored the email service for reuse
- added corresponding validation pipes to two mutations to verify the validity of data input

### Work flow:

- The user enters their email address on the "Forgot Password" page.
- A reset password email should be sent to the user's provided email address (the email contains a link, valid only 1 hours)
- The user clicks the link to navigate to the password reset page and enters a new password.
- The user's password should be update after the link is successfully verified.
- Any errors encountered during any of these steps will return the error code and message.

### Completed:

**Call forgotPassword with 'email' to update reset password token :**
image

**Call resetPassword with the token and the new password:**
image
image

**Reset password token should be deleted after the process:**
image

**Verify the password is updated successfully using Login mutation:**
old password:
image

new password:
image

### Unit test:
image

Resolve CP-28

### CODE REVIEW

File: src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -212,4 +214,96 @@ export class AuthService {
...
+      return {
+        code: RESET_PASSWORD_ERROR,
+        message: 'An unexpected error occurred during sending reset password email.',
+      };
+    }
+  }
+
+  async resetPassword(input: ResetPasswordInput): Promise<Result> {
+    try {
+      const { newPassword, resetToken } = input;
```
Comment (Chris-Wang) [inline src/modules/auth/auth.service.ts:251]: resetToken should not be inside the input, it's used to be gotten from header.

And validate token should be a step prior to resetPassword process.

File: src/common/utils/helpers.ts [hunk 2]
```diff
@@ -82,3 +87,8 @@ export const refSchema = z
   .string()
   .optional()
   .default(process.env.DEFAULT_REFERENCE || 'Default_Reference_Name');
+
+export const resetTokenSchema = z
```
Comment (Chris-Wang) [inline src/common/utils/helpers.ts]: why a token has a schema?

File: src/modules/auth/auth.service.ts [hunk 3]
```diff
@@ -212,4 +214,96 @@ export class AuthService {
...
+          code: RESET_PASSWORD_ERROR,
+          message: "account doesn't exist",
+        };
+      }
+
+      const resetPasswordToken = this.emailVerificationService.generateVerificationToken(email);
+      const res = await this.userService.update(user.id, {
+        resetPasswordToken,
+      });
+      const requestResetUser = await this.userService.findByEmail(email);
```
Comment (Chris-Wang) [inline src/modules/auth/auth.service.ts]: why there is user query again?

**Review by Chris-Wang — CHANGES_REQUESTED** [review]

File: src/common/utils/helpers.ts (hunk 2 above)
Comment (Maaayhan) [reply src/common/utils/helpers.ts]: yeah it's redundant, we'll use jwt function to check the token's validity. I will remove the schema here.

**Review by Chris-Wang — APPROVED** [review]
LGTM

**Pull request merged**